spyder/plugins/console/widgets/main_widget.py

```diff
@@ -0,0 +1,535 @@
+                   args=None):
+        """
+        Run a Python script.
+        """
+        if filename is None:
+            self.shell.interpreter.restore_stds()
+            filename, _selfilter = getopenfilename(
+                self,
+                _("Run Python script"),
+                getcwd_or_home(),
+                _("Python scripts") + " (*.py ; *.pyw ; *.ipy)",
+            )
+            self.shell.interpreter.redirect_stds()
+
+            if filename:
+                os.chdir(osp.dirname(filename))
+                filename = osp.basename(filename)
+            else:
+                return
+
+        logger.debug("Running script with %s", args)
+        filename = osp.abspath(filename)
+        rbs = remove_backslashes
+        command = "runfile('%s', args='%s')" % (rbs(filename), rbs(args))
+
+        if set_focus:
+            self.shell.setFocus()
+
+        self.change_visibility(True, True)
+
+        self.shell.write(command+'\n')
+        self.shell.run_command(command)
+
+    def go_to_error(self, text):
+        """
+        Go to error if relevant.
+        """
+        match = get_error_match(to_text_string(text))
+        if match:
+            fname, lnb = match.groups()
+            self.edit_script(fname, int(lnb))
+
+    def edit_script(self, filename=None, goto=-1):
+        """
+        Edit script.
+        """
+        if filename is not None:
+            # Called from InternalShell
+            self.shell.external_editor(filename, goto)
+            self.sig_edit_goto_requested.emit(osp.abspath(filename), goto, '')
+
+    def execute_lines(self, lines):
+        """
+        Execute lines and give focus to shell.
+        """
+        self.shell.execute_lines(to_text_string(lines))
+        self.shell.setFocus()
+
+    @Slot()
+    def change_max_line_count(self, value=None):
+        """"
+        Change maximum line count.
+        """
+        valid = True
+        if value is None:
+            value, valid = QInputDialog.getInt(
+                self,
+                _('Buffer'),
+                _('Maximum line count'),
+                self.get_option('max_line_count'),
+                0,
+                1000000,
+            )
+
+        if valid:
+            self.set_option('max_line_count', value)
+
+    @Slot()
+    def change_exteditor(self, path=None):
+        """
+        Change external editor path.
+        """
+        valid = True
+        if path is None:
+            path, valid = QInputDialog.getText(
+                self,
+                _('External editor'),
+                _('External editor executable path:'),
+                QLineEdit.Normal,
+                self.get_option('external_editor/path'),
+            )
+
+        if valid:
+            self.set_option('external_editor/path', to_text_string(path))
+
+    # --- API
+    # ------------------------------------------------------------------------

```

Comment:
This comments is probably not necessary because the methods below are part of the public API, right?

Also, all those methods are missing a docstring.